File: src/baekJoon/graph/bfs/불.java
package baekJoon.graph.bfs;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class 불 {

  private static int[] dx = {-1, 1, 0, 0};
  private static int[] dy = {0, 0, -1, 1};
  private static int w, h, answer;
  private static char[][] map;
  private static boolean[][] visited;
  private static Queue<Board> people;
  private static Queue<Board> fires;

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    StringTokenizer st = new StringTokenizer(br.readLine());

    int t = Integer.parseInt(st.nextToken());

    for(int testcase = 0; testcase < t; testcase++) {

      st = new StringTokenizer(br.readLine());
      w = Integer.parseInt(st.nextToken());
      h = Integer.parseInt(st.nextToken());

      map = new char[h][w];
      fires = new LinkedList<>();
      people = new LinkedList<>();
      visited = new boolean[h][w];
      answer = -1;

      for(int i = 0; i < h; i++) {
        st = new StringTokenizer(br.readLine());
        String s = st.nextToken();
        for(int j = 0; j < w; j++) {
          map[i][j] = s.charAt(j);
          visited[i][j] = true;
          if(map[i][j] == '@') {
            people.add(new Board(i, j));
          } else if(map[i][j] == '*') {
            fires.add(new Board(i, j));
          } else if(map[i][j] == '.') {
            visited[i][j] = false;
          }
        }
      }

      sol();
      if(answer == -1) bw.write("IMPOSSIBLE");
      else bw.write(String.valueOf(answer + 1));
      bw.newLine();
    }
    bw.flush();
    bw.close();
  }

  private static void sol() {
    int result = 0;
    while(true) {
      if(answer != -1) break;
      if(people.isEmpty()) break;
      moveFire();
      move(result);
      result++;
    }
  }

  private static void move(int count) {
    int size = people.size();
    for(int i = 0; i < size; i++) {
      if(people.isEmpty()) break;
      Board here = people.poll();
      int cx = here.x;
      int cy = here.y;
      if(isAnswer(cx, cy)) {
        answer = count;
        break;
      }

      for(int j = 0; j < 4; j++) {
        int nx = cx + dx[j];
        int ny = cy + dy[j];
        if(nx < 0 || nx >= h || ny < 0 || ny >= w) continue;
        if(visited[nx][ny]) continue;
        if(isAnswer(nx, ny)) {
          answer = count + 1;
          return;
        }
        visited[nx][ny] = true;
        people.add(new Board(nx, ny));
      }
    }
  }

  private static void moveFire() {
    int size = fires.size();
    for(int i = 0; i < size; i++) {
      if(fires.isEmpty()) break;
      Board fire = fires.poll();
      int fx = fire.x;
      int fy = fire.y;

      for(int j = 0; j < 4; j++) {
        int nx = fx + dx[j];
        int ny = fy + dy[j];
        if(nx < 0 || nx >= h || ny < 0 || ny >= w) continue;
        if(visited[nx][ny]) continue;
        visited[nx][ny] = true;
        fires.add(new Board(nx, ny));
      }
    }
  }

  private static boolean isAnswer(int x, int y) {
    if(x == 0 || x == h - 1) return true;
    return y == 0 || y == w - 1;
  }

  private static class Board {
    int x, y;

    public Board(int x, int y) {
      this.x = x;
      this.y = y;
    }
  }

}

File: src/kotlin/불.kt
import java.io.BufferedWriter
import java.io.OutputStreamWriter
import java.util.*

private val dx = intArrayOf(-1, 1, 0, 0)
private val dy = intArrayOf(0, 0, -1, 1)
private var w = 0
private var h = 0
private var answer = 0
private lateinit var map: Array<CharArray>
private lateinit var visited: Array<BooleanArray>
private lateinit var movedQueue: Queue<Position>
private lateinit var fireQueue: Queue<Position>

fun main() = with(System.`in`.bufferedReader()) {

    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val t = readLine().toInt()

    repeat(t) {

        val (width, height) = readLine().split(" ").map { it.toInt() }
        w = width
        h = height
        answer = -1

        map = Array(h) { CharArray(w) }
        visited = Array(h) { BooleanArray(w) }
        movedQueue = ArrayDeque()
        fireQueue = ArrayDeque()

        for (i in 0 until h) {
            val line = readLine()
            for (j in 0 until w) {
                map[i][j] = line[j]
                visited[i][j] = when(map[i][j]) {
                    '.' -> false
                    '@' -> {
                        movedQueue.add(Position(i, j))
                        false
                    }
                    '*' -> {
                        fireQueue.add(Position(i, j))
                        true
                    }
                    else -> true
                }
            }
        }

        solve()

        if (answer == -1) {
            bw.write("IMPOSSIBLE")
        }
        else {
            bw.write("${answer + 1}")
        }

        bw.newLine()
    }
    bw.flush()
    bw.close()
}

private fun solve() {
    var result = 0
    while (true) {
        if (answer != -1 || movedQueue.isEmpty()) {
            break
        }
        moveFire()
        move(result++)
    }
}

private fun moveFire() {

    repeat(fireQueue.size) {
        val (fx, fy) = fireQueue.poll() ?: return

        for (i in 0 until 4) {
            val nx = fx + dx[i]
            val ny = fy + dy[i]

            if (nx in 0 until h && ny in 0 until w && !visited[nx][ny]) {
                visited[nx][ny] = true
                fireQueue.add(Position(nx, ny))
            }
        }
    }
}

private fun move(count: Int) {

    repeat(movedQueue.size) {

        val (cx, cy) = movedQueue.poll() ?: return
        if (cx == 0 || cy == 0 || cx == h - 1 || cy == w - 1) {
            answer = count
            return
        }

        for (i in 0 until 4) {
            val nx = cx + dx[i]
            val ny = cy + dy[i]

            if (nx in 0 until h && ny in 0 until w && !visited[nx][ny]) {
                visited[nx][ny] = true
                movedQueue.add(Position(nx, ny))
            }
        }
    }
}

data class Position(val x: Int, val y: Int)